packages/snap/src/state.ts

```diff
@@ -0,0 +1,367 @@
+    if (pk.startsWith("0x")) {
+      pk = pk.substring(2);
+    }
+
+    // create the wallet
+    let wallet = await DirectSecp256k1Wallet.fromKey(
+      Uint8Array.from(Buffer.from(pk, "hex")),
+      chain.bech32_prefix
+    );
+
+    let address = (await wallet.getAccounts())[0].address;
+
+    return {
+      address,
+      chain_id
+    };
+  }
+  /**
+   * Gets all Cosmos chains from Metamask snap state.
+   *
+   * @returns All chains in Metamask snap state.
+   * @throws If an error occurs.
+   */
+  public static async getChains(): Promise<Chains> {
+    let data = await snap.request({
+      method: "snap_manageState",
+      params: { operation: "get" },
+    });
+    if (data?.chains == undefined || data?.chains == null) {
+      throw new Error("Snap has not been initialized. Please initialize snap.");
+    }
+    return new Chains(JSON.parse(data?.chains?.toString()!));
+  }
+  /**
+   * Gets a Cosmos chain based on its chain id from Metamask snap state.
+   *
+   * @returns Chain object you want.
+   * @throws If an error occurs.
+   */
+  public static async getChain(chain_id: string): Promise<Chain | null> {
+    const data = await snap.request({
+      method: "snap_manageState",
+      params: { operation: "get" },
+    });
+    if (data?.chains == undefined || data?.chains == null) {
+      throw new Error("Snap has not been initialized. Please initialize snap.");
+    }
+    let chains = new Chains(JSON.parse(data?.chains?.toString()!));
+    let chain = chains.getChain(chain_id);
+    return chain;
+  }
+  /**
+   * Adds a new Cosmos chain into the current Metamask snap state.
+   *
+   * @param chain Chain object to add into state.
+   * @returns The current state of Chains.
+   * @throws If an error occurs.
+   */
+  public static async addChain(chain: Chain): Promise<Chains> {
+    // get the current state of chains in Metamask we will add chain into
+    const data = await snap.request({
+      method: "snap_manageState",
+      params: { operation: "get" },
+    });
+
+    // parse the JSON data into a plain object
+    let parsedData = JSON.parse(data?.chains?.toString()!);
+
+    // create a new Chains object and populate it with the parsed data
+    let chains = new Chains(parsedData);
+
+    // add the chain into chains class
+    chains.addChain(chain);
+
+    // update Metamask state with new chain state
+    await snap.request({
+      method: "snap_manageState",
+      params: { operation: "update", newState: { ...data, chains: chains.string() } },
+    });
+
+    return chains;
+  }
+  /**
+   * Adds a list of Chains as state into Metamask snap.
+   *
+   * WARNING: This replaces the entire state within Metamask so any chains in state will be replaced.
+   *
+   * @param chains Chains to replace entire Metamask chain state with.
+   * @returns Returns the current state of Chains.
+   * @throws If an error occurs.
+   */
+  public static async addChains(chains: Chains): Promise<Chains> {
```

Comment:
I'd make this WARNING clearer by changing the name of this function.
```suggestion
  public static async replaceAllChains(chains: Chains): Promise<Chains> {
```